lib/spack/spack/cmd/__init__.py

```diff
@@ -226,13 +274,98 @@ def disambiguate_spec_from_hashes(spec, hashes, local=False,
             database query. See ``spack.database.Database._query`` for details.
     """
     if local:
-        matching_specs = spack.store.db.query_local(spec, hashes=hashes,
-                                                    installed=installed)
+        pre_matching_specs = spack.store.db.query_local(spec, hashes=hashes,
+                                                        installed=installed)
     else:
-        matching_specs = spack.store.db.query(spec, hashes=hashes,
-                                              installed=installed)
+        pre_matching_specs = spack.store.db.query(spec, hashes=hashes,
+                                                  installed=installed)
+
+    matching_specs = []
+    conflicting_specs = []
+    narrowed_specs = []
+    loaded_specs = _find_already_loaded()
+
+    if compatible:
+        # compute the lists of conflicting and matching specs
+        loaded_hashes = [loaded_specs[name].dag_hash()
+                         for name in loaded_specs]
+        for possible_spec in pre_matching_specs:
+            conflict = False
+            for node in possible_spec.traverse():
+                if (loaded_specs.get(node.name) and
+                    (loaded_specs[node.name].dag_hash() !=
+                     node.dag_hash())):
+                    possible_spec._message = node.name
+                    conflict = True
+                    break
+            if possible_spec.dag_hash() in loaded_hashes:
+                tty.die(str(possible_spec.name) + ' is alreaded loaded\n')
+            if conflict:
+                conflicting_specs.append(possible_spec)
+            else:
+                matching_specs.append(possible_spec)
+    else:
+        matching_specs = pre_matching_specs
+
+    if preferred:
```

Comment:
I think this would be improved by creating a metric for spec "closeness" and sorting by that metric, rather than having a boolean flag for preference.